File: 04/js/04.js
#!/usr/bin/env node

var fs = require('fs');
fs.readFile('input', 'utf-8', function(_, s) {
    let start, stop, part1 = part2 = 0;
    [start, stop] = s.split('-').map(x => +x);
    for (; start <= stop; ++start) {
        let chars = (''+start).split('');
        if (!chars.every((x, i) => !i || chars[i-1] <= x)) continue;
        part1 += /(.)\1/.test(start);
        part2 += /(.)\1/.test((''+start).replace(/(.)\1\1+/g, ''));
    }
    console.log(part1, part2);
});

File: 04/c/04.c
#include <stdio.h>

int main() {
    FILE *fp = fopen("input", "r");
    int i, stop;
    fscanf(fp, "%d-%d", &i, &stop);
    fclose(fp);

    int part1 = 0, part2 = 0;
    for (; i <= stop; ++i) {
        int prev = 10, n, d, p1 = 0, p2 = 0;
        for (int j = i; j; j /= 10) {
            d = j % 10;
            if (d > prev) goto bad;
            if (d == prev) p1 |= ++n == 2;
            else p2 |= n == 2, prev = d, n = 1;
        }
        part1 += p1;
        part2 += p2 | (n == 2);
        bad:{}
    }

    printf("%d %d\n", part1, part2);
}
